### PR TITLE
C++: Disable _some_ constant folding in IR

### PR DESCRIPTION
**Pull request by MathiasVP**
In https://github.com/github/codeql/pull/14210 Alex tried to disable close to all constant folding in IR generation. This made us hit some really complex cases (see the corresponding internal issue), which meant we had to back out of this approach.

This PR does the minimal changes necessary to fix the original issue: being able to select the dataflow node for `1` (or `2`) in an expression such as `1 | 2`. We do this by disabling constant folding _for binary operations only_. In addition, we only disable it for "small binary operations" (currently defined as an operation with less than 50 operands).

Unsurprisingly, most binary operations are "small". Here's the distribution on Wireshark (with a threshold line at 50 to make it easy to spot which ones we fold and which ones we don't):
Screenshot 2024-03-19 at 11 43 18

I imagine we could make this threshold a lot bigger, but this seems to be the safe way to go about this.

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll [hunk 1]
```diff
@@ -40,12 +40,28 @@ IRTempVariable getIRTempVariable(Locatable ast, TempVariableTag tag) {
   result.getTag() = tag
 }
 
+private Expr getAnOperand(BinaryOperation binOp) { result = binOp.getAnOperand() }
+
+private int getNumberOfBinaryOperands(BinaryOperation binOp) {
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll:45]: I expect a binary operation to always have exactly two operands 😅, so the meaning is slightly different here?

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll [hunk 2]
```diff
@@ -40,12 +40,28 @@ IRTempVariable getIRTempVariable(Locatable ast, TempVariableTag tag) {
...
+predicate isIRConstant(Expr expr) {
+  exists(expr.getValue()) and
+  // We avoid constant folding binary operations since it's often useful to
+  // mark one of those as a source in dataflow, and if the operation is
+  // constant folded it's not possible to mark its operands as a source (or
+  // sink).
+  // But to avoid creating an outrageous amount of IR from very large
+  // constant expressions we fall back to constant folding if the operation
+  // has more than 50 operands (i.e., 1 + 2 + 3 + 4 + ... + 50)
+  if expr instanceof BinaryOperation then getNumberOfBinaryOperands(expr) > 50 else any()
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll:63]: I wonder if this should also be limited to just bitwise `&` and `|`?

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll:45]: Oh, good point. I'll add some QLDoc here. The meaning is that `getNumberOfBinaryOperands` grabs all the operations recursively. So `getNumberOfBinaryOperands((1 + 2) + 3)` is `3` even though the outermost binary operation only has two operands (i.e., `1 + 2` and `3`).

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll]: I've added some explanatory QLDoc in a97891cbc131bb7fc94fbede286832a675db2cba, and renamed the predicate to `getNumberOfNestedBinaryOperands` to make it clear that we're not just computing the number 2 😂

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll (hunk 2 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll]: We can do that as an even safer first milestone, sure.

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll]: Fixed in a88d8b260d516ca279ef83cee4d4b4becf662026

**Review by jketema — DISMISSED** [review]
LGTM, if DCA is happy

**Comment by MathiasVP** [comment]
DCA showed three FPs caused by us no longer realizing some code was dead. I've pushed https://github.com/github/codeql/pull/15969/commits/458ee13345a93021f1da581d4cc9ccc36b2918ce to fix this. Locally this fixes those FPs. I'm running another DCA now to confirm this 🤞

**Review by jketema — APPROVED** [review]

**Pull request merged**